src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,316 @@
+            {
+                return;
+            }
+
+            var syntaxContext = await CreateContextAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+            var items = await GetCompletionItemsAsync(document, syntaxContext, position, cancellationToken).ConfigureAwait(false);
+            completionContext.AddItems(items);
+#if DEBUG
+            DebugObject.debug_total_time_with_ItemCreation = System.Environment.TickCount - tick;
+#endif
+            return;
+        }
+
+        private async Task<ImmutableArray<CompletionItem>> GetCompletionItemsAsync(Document document, SyntaxContext context, int position, CancellationToken cancellationToken)
+        {
+            if (context.IsTypeContext)
+            {
+                using (Logger.LogBlock(FunctionId.Completion_TypeImportCompletionProvider_GetCompletionItemsAsync, cancellationToken))
+                {
+                    var project = document.Project;
+                    var node = context.LeftToken.Parent;
+
+                    var namespacesInScope = GetNamespacesInScope(node, context.SemanticModel, cancellationToken);
+                    var builder = ArrayBuilder<TypeImportCompletionItem>.GetInstance();
+
+                    var declarationsInCurrentProject = await _typeImportCompletionService
+                        .GetAccessibleTopLevelTypesFromProjectAsync(project, namespacesInScope, cancellationToken).ConfigureAwait(false);
+                    builder.AddRange(declarationsInCurrentProject);
+
+                    var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                    foreach (var reference in compilation.References)
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+                        var declarationsInReference = ImmutableArray<TypeImportCompletionItem>.Empty;
+                        if (reference is CompilationReference compilationReference)
+                        {
+                            declarationsInReference = await _typeImportCompletionService.GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+                                project.Solution,
+                                compilation,
+                                compilationReference,
+                                namespacesInScope,
+                                cancellationToken).ConfigureAwait(false);
+
+                        }
+                        else if (reference is PortableExecutableReference peReference)
+                        {
+                            declarationsInReference = _typeImportCompletionService.GetAccessibleTopLevelTypesFromPEReference(
+                                project.Solution,
+                                compilation,
+                                peReference,
+                                namespacesInScope,
```

Comment:
what do you need namespacesInScope for?

Reply:
Types in those namespaces would be excluded